### PR TITLE
Skeleton version of the React + Redux app HP-211

### PR DESCRIPTION
**Pull request by NikoHelle**
First version of the example frontend with authentication. Uses keycloak's official javascript library, but supports also oidc-react library used in Profiili UI. 

Support for two clients was added for few reasons:
- wanting to keep same client (oidc-react) that profiili UI uses
- prefer official library (keycloak) 
- use latest stable versions (oidc-react is in beta) with same backend version (keycloak matches server version)
- keycloak is more transparent to the user, but oidc-react has redux clients too and is battle tested.
- client can be changed and swapped at any time when more features are needed

The authentication client offers modern React hooks and context for easy implementation. Demo components included with Redux support.

resolves HP-211

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+staging:
+  variables:
+    DOCKER_IMAGE_NAME: "$CI_PROJECT_NAME-staging"
+  only:
+    refs:
+      - develop
+
+production:
+  variables:
+    DOCKER_IMAGE_NAME: "$CI_PROJECT_NAME-production"
```
Comment (akikoskinen) [inline .gitlab-ci.yml:60]: Is this file needed?

File: src/components/Header.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
+import React, { useState } from 'react';
+import { Navigation } from 'hds-react';
+
+import { useKeycloak } from '../clients/keycloak';
+// import { useTranslation } from 'react-i18next';
```
Comment (akikoskinen) [inline src/components/Header.tsx:5]: Remove commented out code

File: src/components/Header.tsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    { code: 'fi', label: 'Suomi' },
+    { code: 'sv', label: 'Svenska' },
+    { code: 'en', label: 'English' }
+  ];
+  /*
+  const { i18n } = useTranslation();
+  const setLanguage = (code: string) => {
+    i18n.changeLanguage(code);
+  };
+  */
```
Comment (akikoskinen) [inline src/components/Header.tsx:20]: Remove commented out code

File: src/components/Header.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  const [language, setLanguage] = useState(languageOptions[0]);
+  const [active, setActive] = useState<'link' | 'button' | 'dropdown'>();
+
+  // show helsingfors logo if swedish is selected as the language
+  const logoLanguage = language.code === 'sv' ? 'sv' : 'fi';
+
+  const title =
+    language.code === 'sv'
+      ? 'Helsingfors Systembolaget'
+      : 'Helsinki Design System';
```
Comment (akikoskinen) [inline src/components/Header.tsx:35]: Just for the looks, maybe use proper titles?

File: src/i18n/en.json [hunk 5]
```diff
@@ -0,0 +1,160 @@
+{
+  "appName": "Profile",
+  "berthErrors": {
+    "close": "Close",
+    "explanation": "You cannot delete your information because Boat berths service needs your information for billing. Contact the customer service of Berth service to terminate your contract. After that has been done, you can try again.",
+    "title": "The operation cannot be completed."
+  },
```
Comment (akikoskinen) [inline src/i18n/en.json:7]: This file contains some texts that probably aren't used in this demo.

File: src/i18n/fi.json [hunk 6]
```diff
@@ -0,0 +1,160 @@
+{
+  "appName": "Profiili",
+  "berthErrors": {
+    "close": "Sulje",
+    "explanation": "Et voi poistaa tietojasi, koska Venepaikka-palvelu tarvitsee tietojasi laskutukseen. Ota yhteyttä Venepaikka-palvelun asiakaspalveluun irtisanoaksesi sopimuksesi ja yritä sen jälkeen poistoa uudelleen.",
+    "title": "Toimintoa ei voi suorittaa."
+  },
```
Comment (akikoskinen) [inline src/i18n/fi.json:7]: This file contains some texts that probably aren't used in this demo.

File: src/i18n/sv.json [hunk 7]
```diff
@@ -0,0 +1,160 @@
+{
+  "appName": "Profil",
+  "berthErrors": {
+    "close": "Stänga",
+    "explanation": "Du kan inte radera din information eftersom Båtplatser-tjänsten behöver din faktureringsinformation. Kontakta Båtplatsers kundservice för att säga upp ditt kontrakt. När det är gjort kan du försöka igen.",
+    "title": "Operationen kan inte slutföras."
+  },
```
Comment (akikoskinen) [inline src/i18n/sv.json:7]: This file contains some texts that probably aren't used in this demo.

File: src/setupTests.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+customGlobal.fetchMock = customGlobal.fetch;
+
+configure({ adapter: new Adapter() });
+
+jest.mock('react-router', () => ({
+  ...jest.requireActual('react-router'),
+  useHistory: (): Record<string, Function> => ({
+    push: jest.fn()
+  })
+}));
```
Comment (akikoskinen) [inline src/setupTests.ts:20]: What are test related files needed for, if there aren't any tests?

**Review by akikoskinen — CHANGES_REQUESTED** [review]
I don't especially like the way how these commits first bring in a ton of files in earlier commits and then remove quite a bunch of them in later commits. It pollutes the commit history and, as this is supposed to be an example, IMO doesn't work as a good example in that way. So I would **like** to see the commits cleaned. But I also understand, if that would be seen as too little gain for too much work, so I won't be requiring it.

Since the app isn't really translated, perhaps just remove everything related to translations?

Same with tests: as there aren't any real tests, remove everything related to tests?

**Comment by akikoskinen** [comment]
When signing in (for the first time I guess), this is shown: 

Screenshot 2020-09-25 at 9 54 10

Why is the "Example UI Profile" line in the list of access privileges to be granted?

The page is in Tunnistamo, but I guess the Example UI somehow provides that list of items it wants to access.

File: .travis.yml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  include:
+    - stage: 'Test'
+      name: 'Run code linting and test'
+      script: yarn lint && yarn ci
+    - stage: 'Build'
+      name: 'Run (test) build'
+      if: branch IN (master, develop)
+      script: yarn build
+after_success:
+  - ./node_modules/.bin/codecov < coverage/lcov.info
```
[the remark this answers is not in the document]
Comment (NikoHelle) [reply .travis.yml]: Removed.

File: .vscode/settings.json [hunk 10]
```diff
@@ -0,0 +1,10 @@
+{
+  "eslint.validate": [
+    "javascript",
+    "javascriptreact",
+    { "language": "typescript", "autoFix": true },
+    { "language": "typescriptreact", "autoFix": true }
+  ],
+  "typescript.tsdk": "node_modules/typescript/lib",
+  "workbench.editor.enablePreview": false
+}
```
[the remark this answers is not in the document]
Comment (NikoHelle) [reply .vscode/settings.json]: Removed.

File: README.md [hunk 11]
```diff
@@ -1,2 +1,38 @@
 # example-profile-ui
-Example UI application that interacts with Helsinki profile
+Example UI application that interacts with Keycloak (and later also with Helsinki profile)
+
+There are two clients that can be used for authentication: keycloak.js and oidc-react.js. The version of the oidc-react.js npm library is beta and keycloak is the official package released with keycloak server. Currently keycload client is used in this demo.
+
+Included in this demo app:
+- two login clients
```
[the remark this answers is not in the document]
Comment (NikoHelle) [reply README.md:7]: Added!

File: README.md [hunk 12]
```diff
@@ -1,2 +1,38 @@
...
+- HOC component listening a client and showing different content for authorized and unauthorized users.
+
+Clients dispatch events and trigger changes which then trigger re-rendering of the components using clients.
+
+
+## Config
+use .env -files. Some values are client specific. Default client is keycloak and relevant settings are:
+REACT_APP_OIDC_URL="https://tunnistus.hel.ninja/auth"
+REACT_APP_OIDC_REALM="helsinki-tunnistus"
+REACT_APP_OIDC_CLIENT_ID="https://api.hel.fi/auth/example-ui-profile"
```
[the remark this answers is not in the document]
Comment (NikoHelle) [reply README.md:28]: Done

File: README.md [hunk 13]
```diff
@@ -1,2 +1,38 @@
...
+
+Starting server with temporary environment variables:
+REACT_APP_OIDC_URL=https://foo.bar yarn start
+
+## Docker
+
+Run `docker-compose up`
+
+Starting docker with temporary environment variables:
+Open docker-compose.yml and add new new 'environment' under services/app.
```
[the remark this answers is not in the document]
Comment (NikoHelle) [reply README.md]: Fixed.

File: README.md [hunk 14]
```diff
@@ -1,2 +1,38 @@
...
+
+Example:
+
+services:
+  app:
+    environment: 
+      - REACT_APP_OIDC_URL=https://foo.bar
```
[the remark this answers is not in the document]
Comment (NikoHelle) [reply README.md]: Done.

File: scripts/update-translations.ts [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+  try {
+    await fetchTranslations(sheetId, languages, pathToLocales);
+    console.log('Done'); // eslint-disable-line
+  } catch (err) {
+    console.error(err.message); // eslint-disable-line
+    process.exit(1);
+  }
+};
+
+start();
```
[the remark this answers is not in the document]
Comment (NikoHelle) [reply scripts/update-translations.ts:22]: Removed all i18n stuff

File: src/clients/index.ts [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+  flow: 'standard' | 'implicit' | 'hybrid' | undefined;
+}
+
+type EventHandlers = {
+  addListener: Client['addListener'];
+  eventTrigger: (eventType: ClientEventIds, payload?: EventPayload) => void;
+};
+
+export const createEventHandling = (): EventHandlers => {
+  const listeners: Map<ClientEventIds, Set<EventListener>> = new Map(); // todo:
```
[the remark this answers is not in the document]
Comment (NikoHelle) [reply src/clients/index.ts]: :) Cleared.

File: src/clients/keycloak.ts [hunk 17]
```diff
@@ -0,0 +1,310 @@
...
+
+function setSessionStorageTokens({
+  token,
+  idToken,
+  refreshToken
+}: {
+  token: string | undefined;
+  idToken: string | undefined;
+  refreshToken: string | undefined;
+}): void {
```
[the remark this answers is not in the document]
Comment (NikoHelle) [reply src/clients/keycloak.ts]: Done.

File: src/clients/keycloak.ts [hunk 18]
```diff
@@ -0,0 +1,310 @@
...
+  localStorage.setItem('token', token || '');
+  localStorage.setItem('idToken', idToken || '');
+  localStorage.setItem('refreshToken', refreshToken || '');
+}
+
+function getSessionStorageTokens(): {
+  token: string | undefined;
+  idToken: string | undefined;
+  refreshToken: string | undefined;
+} {
```
[the remark this answers is not in the document]
Comment (NikoHelle) [reply src/clients/keycloak.ts]: Done.

File: src/components/Header.tsx (hunk 2 above)
Comment (NikoHelle) [reply src/components/Header.tsx]: Fixed.

File: src/components/Header.tsx (hunk 3 above)
Comment (NikoHelle) [reply src/components/Header.tsx]: Done.

File: src/components/Header.tsx (hunk 4 above)
Comment (NikoHelle) [reply src/components/Header.tsx]: Removed.

File: src/i18n/fi.json (hunk 6 above)
Comment (NikoHelle) [reply src/i18n/fi.json]: File removed.

File: src/i18n/sv.json (hunk 7 above)
Comment (NikoHelle) [reply src/i18n/sv.json]: File removed.

File: src/setupTests.ts (hunk 8 above)
Comment (NikoHelle) [reply src/setupTests.ts:23]: Tests added

File: src/clients/index.ts [hunk 19]
```diff
@@ -32,7 +33,7 @@ export const ClientStatus = {
   UNAUTHORIZED: 'UNAUTHORIZED'
 } as const;
 
-export type ClientStatusIds = typeof ClientStatus[keyof typeof ClientStatus]; // todo change plural -> single
+export type ClientStatusId = typeof ClientStatus[keyof typeof ClientStatus]; // todo change plural -> single
```
Comment (akikoskinen) [inline src/clients/index.ts]: Is the `todo` text still applicable?

File: src/clients/index.ts [hunk 20]
```diff
@@ -44,7 +45,7 @@ export const ClientEvent = {
   ...ClientStatus
 } as const;
 
-export type ClientEventIds = typeof ClientEvent[keyof typeof ClientEvent]; // todo change plural -> single
+export type ClientEventId = typeof ClientEvent[keyof typeof ClientEvent]; // todo change plural -> single
```
Comment (akikoskinen) [inline src/clients/index.ts]: Is the `todo` text still applicable?

File: src/clients/__tests__/index.ts [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+  setTokens: (newTokens: {}) => {};
+  getInstance: () => Keycloak.KeycloakInstance;
+  createValidUserData: (props?: {}) => { email: string; session_state: string };
+};
+
+describe('File must return one test ', () => {
+  it(' and this is just a note:', () => {
+    expect(!!'skip').toBeTruthy();
+  });
+});
```
Comment (akikoskinen) [inline src/clients/__tests__/index.ts:81]: Is this kind of stuff **really** needed? Looks pretty 💩

File: src/clients/__tests__/oidc-react.test.ts [hunk 22]
```diff
@@ -0,0 +1,9 @@
+describe('File must return one test ', () => {
+  it(' and this is just a note:', () => {
+    expect(
+      !!'not done, as keycloak will likely be the chosen client'
+    ).toBeTruthy();
+  });
+});
+
+export default 'this line prevents lint error';
```
Comment (akikoskinen) [inline src/clients/__tests__/oidc-react.test.ts:9]: Looks like a pretty useless file to me.

File: src/clients/index.ts [hunk 23]
```diff
@@ -33,7 +34,7 @@ export const ClientStatus = {
   UNAUTHORIZED: 'UNAUTHORIZED'
 } as const;
 
-export type ClientStatusId = typeof ClientStatus[keyof typeof ClientStatus]; // todo change plural -> single
```
Comment (akikoskinen) [inline src/clients/index.ts]: Aha, these obsolete `todo`s are removed in another commit. Would really like to see these changes done in the commits where they nicely fit.

File: src/AppMeta.tsx [hunk 24]
```diff
@@ -1,15 +0,0 @@
-import React from 'react';
-import { Helmet } from 'react-helmet';
```
Comment (akikoskinen) [inline src/AppMeta.tsx]: Can `react-helmet` now be removed from the project?

File: src/AppMeta.tsx [hunk 25]
```diff
@@ -1,15 +0,0 @@
...
-import { useTranslation } from 'react-i18next';
```
Comment (akikoskinen) [inline src/AppMeta.tsx]: ...and `react-i18next` as well?

**Review by akikoskinen — COMMENTED** [review]
Some findings...

File: src/AppMeta.tsx (hunk 25 above)
Comment (NikoHelle) [reply src/AppMeta.tsx]: Has been removed.

File: src/AppMeta.tsx (hunk 24 above)
Comment (NikoHelle) [reply src/AppMeta.tsx]: Has been removed.

File: src/clients/__tests__/oidc-react.test.ts (hunk 22 above)
Comment (NikoHelle) [reply src/clients/__tests__/oidc-react.test.ts]: Added actual tests.

File: src/clients/__tests__/index.ts (hunk 21 above)
Comment (NikoHelle) [reply src/clients/__tests__/index.ts]: Jest test file filter cannot be changed via package.json, so made a new __mocks__-folder for these. Same logic is in profile-ui

File: src/AppMeta.tsx (hunk 24 above)
Comment (akikoskinen) [reply src/AppMeta.tsx]: But it's still [there](https://github.com/City-of-Helsinki/example-ui-profile/pull/1/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R42)?

File: src/AppMeta.tsx (hunk 25 above)
Comment (akikoskinen) [reply src/AppMeta.tsx]: But it's still [there](https://github.com/City-of-Helsinki/example-ui-profile/pull/1/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R43)?

File: src/clients/__tests__/client.test.ts [hunk 26]
```diff
@@ -0,0 +1,281 @@
...
+  Client,
+  ClientEvent,
+  ClientError,
+  ClientType
+} from '../index';
+import { createKeycloakClient } from '../keycloak';
+import { mockMutatorGetter } from '../__mocks__/keycloak-mock';
+import { createOidcClient } from '../oidc-react';
+import { mockMutatorGetterOidc } from '../__mocks__/oidc-react-mock';
+// import { createOidcClient } from '../oidc-react';
```
Comment (akikoskinen) [inline src/clients/__tests__/client.test.ts:19]: Remove commented out code.

File: src/clients/__tests__/client.test.ts [hunk 27]
```diff
@@ -0,0 +1,281 @@
...
+          token: 'token',
+          idToken: 'idToken',
+          refreshToken: 'refreshToken'
+        };
+        it('login call is passed to the client library and tokens are saved', async () => {
+          mockMutator.setTokens(tokens);
+          await to(client.init());
+          client.login();
+          expect(mockMutator.getLoginCallCount()).toBe(1);
+          // expect(getSessionStorageTokens()).toEqual(tokens);
```
Comment (akikoskinen) [inline src/clients/__tests__/client.test.ts:187]: Remove commented out code

File: src/clients/__tests__/client.test.ts [hunk 28]
```diff
@@ -0,0 +1,281 @@
...
+        });
+        it('logout call is passed to the client library and event is triggered and tokens are cleared', async () => {
+          mockMutator.setTokens(tokens);
+          await to(client.init());
+          // expect(getSessionStorageTokens()).toEqual(tokens);
```
Comment (akikoskinen) [inline src/clients/__tests__/client.test.ts:192]: Remove commented out code

File: src/clients/__tests__/client.test.ts [hunk 29]
```diff
@@ -0,0 +1,281 @@
...
+          client.logout();
+          expect(mockMutator.getLogoutCallCount()).toBe(1);
+          expect(eventListeners.getCallCount(ClientEvent.LOGGING_OUT)).toBe(1);
+          /* expect(getSessionStorageTokens()).toEqual({
+            token: undefined,
+            idToken: undefined,
+            refreshToken: undefined
+          }); */
```
Comment (akikoskinen) [inline src/clients/__tests__/client.test.ts:200]: Remove commented out code

File: src/clients/__tests__/oidc-react.test.ts [hunk 30]
```diff
@@ -0,0 +1,178 @@
...
+// Inspired by https://blog.grossman.io/how-to-write-async-await-without-try-catch-blocks-in-javascript/
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+async function to(promise: Promise<unknown>) {
+  try {
+    const data = await promise;
+    return [null, data];
+  } catch (err) {
+    return [err];
+  }
+}
```
Comment (akikoskinen) [inline src/clients/__tests__/oidc-react.test.ts:23]: This code seems to be repeated in multiple files. Could it be extracted to a single place somewhere?

File: src/components/Header.tsx [hunk 31]
```diff
@@ -0,0 +1,122 @@
...
+  const [active, setActive] = useState<'link' | 'button' | 'dropdown'>();
+
+  // show helsingfors logo if swedish is selected as the language
+  const logoLanguage = language.code === 'sv' ? 'sv' : 'fi';
+
+  const title = 'Helsinki Profiili Example';
+
+  // formats the selected value
+  const formatSelectedValue = ({ code }: LanguageOption): string =>
+    code.toUpperCase();
```
Comment (akikoskinen) [inline src/components/Header.tsx:30]: There are a few occurrences in this file where a comment repeats exactly the next line, without giving any additional information. IMO that's useless, I would remove such comments.

File: src/setupTests.ts [hunk 32]
```diff
@@ -0,0 +1,61 @@
...
+    return clientInstance;
+  };
+});
+
+jest.mock('oidc-client', () => {
+  class MockUserManagerClass {
+    constructor(settings: UserManagerSettings) {
+      const mockMutator = mockMutatorGetterOidc();
+      const userManager = mockOidcUserManager(settings) as UserManager;
+      // console.log('userManager', userManager);
```
Comment (akikoskinen) [inline src/setupTests.ts:52]: Remove commented out code

**Review by akikoskinen — CHANGES_REQUESTED** [review]
Getting there I guess

File: src/AppMeta.tsx (hunk 24 above)
Comment (NikoHelle) [reply src/AppMeta.tsx]: AppMeta.tsx has been deleted some time ago.

File: src/AppMeta.tsx (hunk 25 above)
Comment (NikoHelle) [reply src/AppMeta.tsx]: AppMeta.tsx has been deleted some time ago.

File: src/clients/__tests__/client.test.ts (hunk 26 above)
Comment (NikoHelle) [reply src/clients/__tests__/client.test.ts]: Done

File: src/clients/__tests__/client.test.ts (hunk 28 above)
Comment (NikoHelle) [reply src/clients/__tests__/client.test.ts]: Done

File: src/clients/__tests__/client.test.ts (hunk 29 above)
Comment (NikoHelle) [reply src/clients/__tests__/client.test.ts]: Done

File: src/clients/__tests__/oidc-react.test.ts (hunk 30 above)
Comment (NikoHelle) [reply src/clients/__tests__/oidc-react.test.ts]: Ah, forgot about this common function. I imported a npm package that does the same.

File: src/components/Header.tsx (hunk 31 above)
Comment (NikoHelle) [reply src/components/Header.tsx]: I didnt like 'em either. :) Came with the file from profile-ui. Removed.

File: src/setupTests.ts (hunk 32 above)
Comment (NikoHelle) [reply src/setupTests.ts]: Done.

File: src/AppMeta.tsx (hunk 24 above)
Comment (akikoskinen) [reply src/AppMeta.tsx]: I mean that `react-helmet` is still in `package.json` but can it be removed now?

File: src/AppMeta.tsx (hunk 25 above)
Comment (akikoskinen) [reply src/AppMeta.tsx]: I mean that `react-i18next` is still in `package.json` but can it be removed now?

**Review by akikoskinen — APPROVED** [review]
With my limited expertise on this type of code, I'm fine with this.
